Terminal In-Person Payments auto-renewing subscription support

joshheald opened the pull request:
Fixes #

#### Changes proposed in this Pull Request

- Add tokens to order and user when collecting a terminal payment for an order containing subscription products.
- Undeprecate `payments/orders/{order_id}/create_customer` endpoint for use in taking payments for subscriptions.

<!--
Title: A descriptive, yet concise, title.
-->
## Experimental Terminal subscription support
<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

When we take payment for an order with subscriptions using the IPP terminal, the subscription renewals are not set up.

This PR address that need, by saving the tokens and preparing the subscription for auto-collection, where we are capable of doing that (country limitations apply.)

[This approach uses PaymentIntents which are set up with `setupFutureUsage = "off_session"` ](https://docs.stripe.com/terminal/features/saving-cards/save-after-payment?terminal-sdk-platform=ios#client-side).

The clients (e.g. mobile apps) are responsible for setting up the subscription, setting up the Customer for the order, and then calling the (now-undeprecated) `create_customer` endpoint to fetch the customer ID. They will attach this to the payment method, along with the setup future usage parameter, and then collect the payment method.

This is submitted as usual to `capture_terminal_payment`.

The changes to the `capture_terminal_payment` endpoint check for any subscriptions related to the order, and for a `generated_card` payment method to use for renewals on the captured payment intent. If both of these are present, it sets up the renewals by adding the reusable payment method to the user, and to each subscription relating to the order, and finally disables manual renewal (where not compulsory.)

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

The code relies on WCS being present, and checks for each function it uses before doing any work, so we don't "half-activate" a subscription if functions are not available.

If the `generated_card` is not present, the endpoint succeeds anyway, even if there are subscriptions on the order. This is because the initial payment will still have been taken, and Subscriptions will handle required manual renewals without risk of the merchant losing money. We could consider adding to the success response in this case so that the apps can alert the merchant that renewals won't be set up until another payment method is provided.

A thought – We could consider adding a pre-capture check for orders with subscription products, but no subscriptions created, and error if that occurs... so we don't capture payments when something has gone wrong with the previous step. I'm not sure whether that would be dealing with the concerns of subscriptions too much, but I suppose the check could be calling another function in WCS to see whether the order is payable yet...?

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
Set up the iOS app (in a simulator if needed) https://github.com/woocommerce/woocommerce-ios/pull/12382 – this allows payment for subscriptions, sets `off_session` future usage, and has somewhere to put the customer ID.

* Add an order in WP admin containing one or more subscription products, with a user selected for the customer.
* Make a POST request to `orders/{order_id}/subscriptions` to create the subscription(s)
* Make a POST request to `payments/orders/112/create_customer` to create the customer in Stripe
* Modify the app code to add the customer ID from the previous step. In [`PaymentIntentParameters.build(_:with:)`](https://github.com/woocommerce/woocommerce-ios/blob/62e1a86654af8d8576f7d2f4e9f28ec059adf925/Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters%2BStripe.swift#L39), you'll see there's a `cus_xyz` ID hardcoded – replace this with the ID from the previous step.
* Set the `-simulate-stripe-card-reader` option in the app build scheme
* Build and run
* Go to the Orders tab in the app
* Select the order
* Tap `Collect Payment`
* Tap `Card`
* Wait for the card payment to be collected
* Check back in the Subscriptions page on WC admin – observe that the subscription has been activated, and in the database see that the payment method has been set.

https://github.com/Automattic/woocommerce-payments/assets/2472348/a1a933bc-e6ff-4370-9c84-b192d06cb388


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

joshheald committed:
Add token service to Orders REST for terminal subs
Supporting subscriptions paid for from the terminal will require that we work with tokens in the `capture_terminal_intent` endpoint.

Here we prepare for that by adding the token service to the Orders REST controller and associated files.

joshheald committed:
Save tokens on terminal payment for a subscription

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8390` or branch name `experiment/hack-week-terminal-subs-support` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=experiment/hack-week-terminal-subs-support) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 972c278a9fb58d1eaef9665257f628a7f96be3cc
- Build time: 2024-04-26 12:32:35 UTC

_Note: the build is updated when a new commit is pushed to this PR._

joshheald committed:
Hacky update of the subs payment method type

joshheald committed:
Use the generated_card for subscriptions

`setup_future_use` means that the returned payment intent has an additional payment method, associated via the charge object. This is under `generated_card`.

The `generated_card` is a `card` payment method, not `card_present`, and it's associated with the customer.

Storing this one allows us to properly charge the customer outside of the initial transaction.

joshheald committed:
Set WCPay gateway on subscriptions on IPP payment

If we have a generated_card payment method to add to the subscription, we need to change it to non-manual renewals, otherwise they won’t be processed automatically.

joshheald committed:
Add tests and safety checks to IPP subscriptions

joshheald committed:
Fix linting errors

joshheald committed:
Add changelog

joshheald committed:
Pre-code review tidy

joshheald committed:
Correctly mock manual_payment_required method

joshheald committed:
Tidy up subscription related function checks

joshheald committed:
Test for correct is_manual value after IPP capture

joshheald committed:
Merge branch 'develop' into experiment/hack-week-terminal-subs-support

joshheald committed:
Undeprecate create_customer endpoint

This endpoint is required for subscriptions to be paid for with IPP.

joshheald committed:
Update tests to remove deprecation expectations

joshheald committed:
Add changelog for endpoint undeprecation

kalessil committed:
Merge branch 'develop' into experiment/hack-week-terminal-subs-support

kalessil committed:
Code review: keep single changelog entry.

kalessil committed:
Code review: changelog entry wording tweaks.

kalessil committed:
Code review: tweak workflows a bit to delay slow functions invocation.

kalessil committed:
Code review: skip a todo.

kalessil committed:
Code review: update psalm baseline.

kalessil reviewed (commented):
Great chunk of work here! LGTM!

PS: I pushed some minor tweaks (arranging nested ifs for better performance - `function_exists` is pretty slow and making CI happy), worth taking a critical look.

kalessil commented:
@Automattic/gamma: surfacing this PR for visibility.

This PR is a result of the last Hack Week and aims to add subscription support to mobile payments.
I see zero risks for WooPayments release (endpoint used by mobile app only) and no backward compatibility risks for older Mobile apps.

I'm adding ya folks as reviewers because this is IPP-related and falls under the team's product areas.

One thing worth clarifying is the urgency of this PR for Mobile teams; @joshheald, can you share insights, please?

zmaglica commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -217,6 +226,30 @@ public function capture_terminal_payment( WP_REST_Request $request ) {
 			}
 			// Store receipt generation URL for mobile applications in order meta-data.
 			$order->add_meta_data( 'receipt_url', get_rest_url( null, sprintf( '%s/payments/readers/receipts/%s', $this->namespace, $intent->get_id() ) ) );
+
+			// Add payment method for future subscription payments.
+			$generated_card = $intent->get_charge()->get_payment_method_details()[ Payment_Method::CARD_PRESENT ]['generated_card'] ?? null;
+			// If we don't get a generated card, e.g. because a digital wallet was used, we can still return that the initial payment was successful.
+			// The subscription will not be activated and customers will need to provide a new payment method for renewals.
+			if ( $generated_card ) {
+				$has_subscriptions = function_exists( 'wcs_order_contains_subscription' ) &&
+										function_exists( 'wcs_get_subscriptions_for_order' ) &&
+										function_exists( 'wcs_is_manual_renewal_required' ) &&
+										wcs_order_contains_subscription( $order_id );
+				if ( $has_subscriptions ) {
+					$token = $this->token_service->add_payment_method_to_user( $generated_card, $order->get_user() );
+					$this->gateway->add_token_to_order( $order, $token );
+					foreach ( wcs_get_subscriptions_for_order( $order ) as $subscription ) {
+						$subscription->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
+						// Where the setting doesn't force manual renewals, we should turn them off, because we have an auto-renewal token now.
+						if ( ! wcs_is_manual_renewal_required() ) {
+							$subscription->set_requires_manual_renewal( false );
+						}
+						$subscription->save();
+					}
+				}
+			}
+
```
Nothing wrong with the code itself, but I would place it outside of the controller. Perhaps we can move it into the order service, and do the checks inside the `mark_terminal_payment_completed` or somewhere similar.

I do understand that this was a hack project, so feel free to threat this comment as a nitpick.

joshheald replied:
Thanks for your review @zmaglica. For context, I work as a mobile engineer and I'm finding my feet with PHP when it makes sense... so apologies if some of my questions are a bit basic!

If we move the code to the order service, we'll add a dependency between the order and token services. Do we want that? The service would also need the gateway in order to add the token. Perhaps we could pass the token creation/addition in a closure to avoid this coupling...

However, this code is quite specific to part of the API used by the mobile app – when the same thing happens on the web, it's handled through the cart and checkout by the subscriptions plugin. It seems unlikely we'd want to use this outside of the REST controller in future. Given that this code "fills a gap" left in the API by the subscriptions plugin, perhaps the REST controller is the right place for it... WDYT?

I'll experiment a little with the changes anyway.

> One thing worth clarifying is the urgency of this PR for Mobile teams; @joshheald, can you share insights, please?
> – https://github.com/Automattic/woocommerce-payments/pull/8390#issuecomment-2079317192

My team will soon be shifting focus to a different project with some urgency, and may hand responsibility for this project over to another team when that happens. In order to do that smoothly, it would be really helpful to merge this sooner rather than later. If you don't have major concerns, perhaps we could continue and improve it in future should the need arise?

gpressutto5 commented:
Adding arguments to an existing controller class causes a fatal error when upgrading the plugin #7464

p1715096110045039/1713879105.398119-slack-C011ENB20Q1

![image](https://github.com/Automattic/woocommerce-payments/assets/12385501/d514610b-99d3-4ba6-a6e9-d93c165d892b)
